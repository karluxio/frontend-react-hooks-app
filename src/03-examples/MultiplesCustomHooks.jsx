import { useCounter, useFetch } from "../hooks"
import { LoadingQuote, Quote } from "./"

export const MultiplesCustomHooks = () => {

  const { counter, increment } = useCounter(1)

  const { data, isLoading, hasError } = useFetch(`https://api.breakingbadquotes.xyz/v1/quotes/${counter}`)


  const { quote, author } = !!data && data[0]

  return (
    <>
      <h1>BreakingBad quotes</h1>

      <hr />

      {hasError && <h3>Error</h3>}

      {
        isLoading
          ? <LoadingQuote />
          : <Quote quote={quote} author={author} />
      }

      <button
        className="btn btn-primary"
        onClick={() => increment(1)}
        disabled={isLoading}>
        Next quote
      </button>
    </>
  )
}