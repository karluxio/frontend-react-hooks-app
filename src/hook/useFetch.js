import { useState, useEffect } from "react"

export const useFetch = (url) => {

  const [state, setState] = useState({
    data: null,
    isLoading: true,
    hasError: null
  })

  useEffect(() => {
    const getGifs = async () => {
      fetch(url)
        .then(resp => resp.json())
        .then(data => setState({ data, loading: false, hasError: null }))
        .catch(err => setState({ data: null, isLoading: false, hasError: err }))
    }

    getGifs()

  }, [url])


  return {
    data: state.data,
    hasError: state.hasError,
    isLoading: state.isLoading,
  }
}