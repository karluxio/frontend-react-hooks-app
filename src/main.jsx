// import React from 'react'
import ReactDOM from 'react-dom/client'

// import { HooksApp } from './HooksApp'
// import { CounterApp } from './01-useState/CounterApp'
// import { CounterWithCustomHook } from './01-useState/CounterWithCustomHook'
// import { SimpleForm } from './02-useEffect/SimpleForm'
// import { FormWithCustomForm } from './02-useEffect/FormWithCustomHook'
// import { MultiplesCustomHooks } from './03-examples/MultiplesCustomHooks'
// import { FocusScreen } from './04-useRef/FocusScreen'
// import { Layout } from './05-useLayoutEffect/Layout'
// import { Memorize } from './06-memos/Memorize'
// import { MemoHook } from './06-memos/MemoHook'
// import { CallbackHook } from './06-memos/CallbackHook'
// import { Padre } from './07-tarea-memo/Padre'

// import './08-useReducer/intro-reducer'

import './index.css'
import { TodoApp } from './08-useReducer/TodoApp'

ReactDOM.createRoot(document.getElementById('root')).render(
  //   // <React.StrictMode>
  <>
    {/* <HooksApp /> */}
    {/* <CounterApp />*/}
    {/* <CounterWithCustomHook /> */}
    {/* <SimpleForm /> */}
    {/* <FormWithCustomForm /> */}
    {/* <MultiplesCustomHooks /> */}
    {/* <FocusScreen /> */}
    {/* <Layout /> */}
    {/* <Memorize /> */}
    {/* <MemoHook /> */}
    {/* <CallbackHook /> */}
    {/* <Padre /> */}
    <TodoApp />
  </>


  // </React.StrictMode>,
)
