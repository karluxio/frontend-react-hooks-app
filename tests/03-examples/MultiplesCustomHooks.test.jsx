import { render, screen, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

import { MultiplesCustomHooks } from "../../src/03-examples"
import { useFetch } from '../../src/hooks/useFetch.js'
import { useCounter } from '../../src/hooks/useCounter.js'

jest.mock('../../src/hooks/useFetch.js')
jest.mock('../../src/hooks/useCounter.js')

describe('MultiplesCustomHooks', () => {

  beforeEach(() => {
    jest.clearAllMocks()
  })

  const mockIncrement = jest.fn()

  useCounter.mockReturnValue({
    counter: 1,
    increment: mockIncrement,

  })

  it("should match against snapshot", () => {

    useFetch.mockReturnValue({ data: null, isLoading: true, hasError: null })

    const { container } = render(<MultiplesCustomHooks />)

    expect(container).toMatchSnapshot()
  })

  it("should return a default component", () => {

    useFetch.mockReturnValue({ data: null, isLoading: true, hasError: null })

    render(<MultiplesCustomHooks />)

    expect(screen.getByText("BreakingBad quotes"))
    expect(screen.getByText("Loading..."))

    const nextButton = screen.getByRole("button", { name: /next/i })

    expect(nextButton).toBeInTheDocument()

    expect(nextButton.disabled).toBeTruthy()

    // screen.debug()
  })

  it("should render a quote", () => {


    useFetch.mockReturnValue({
      data: [{ author: 'john doe', quote: 'some quote' }],
      isLoading: false,
      hasError: null
    })

    render(<MultiplesCustomHooks />)

    expect(screen.getByText(/some quote/i)).toBeTruthy()
    expect(screen.getByText(/john doe/i)).toBeTruthy()

    const nextButton = screen.getByRole("button", { name: /next/i })
    expect(nextButton.disabled).toBeFalsy()

    // screen.debug()

  })

  it("should call increment function", () => {

    useFetch.mockReturnValue({
      data: [{ author: 'john doe', quote: 'some quote' }],
      isLoading: false,
      hasError: null
    })

    render(<MultiplesCustomHooks />)

    const nextButton = screen.getByRole("button", { name: /next/i })
    fireEvent.click(nextButton)

    expect(useCounter).toHaveBeenCalled()
  })
})