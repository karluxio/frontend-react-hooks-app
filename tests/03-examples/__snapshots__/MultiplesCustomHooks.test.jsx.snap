// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiplesCustomHooks should match against snapshot 1`] = `
<div>
  <h1>
    BreakingBad quotes
  </h1>
  <hr />
  <div
    class="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    class="btn btn-primary"
    disabled=""
  >
    Next to quote 
    2
  </button>
</div>
`;
